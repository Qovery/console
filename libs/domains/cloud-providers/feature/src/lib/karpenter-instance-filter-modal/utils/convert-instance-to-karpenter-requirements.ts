import {
  type ClusterInstanceTypeResponseListResultsInner,
  type KarpenterNodePoolRequirement,
} from 'qovery-typescript-axios'

export function convertToKarpenterRequirements(
  instance: ClusterInstanceTypeResponseListResultsInner[]
): KarpenterNodePoolRequirement[] {
  if (!instance || instance.length === 0) {
    return []
  }

  const requirements: KarpenterNodePoolRequirement[] = []

  const uniqueValues = {
    architectures: new Set<string>(),
    families: new Set<string>(),
    sizes: new Set<string>(),
  }

  instance.forEach((instance) => {
    if (instance.architecture) {
      uniqueValues.architectures.add(instance.architecture)
    }

    if (instance.attributes?.instance_family) {
      uniqueValues.families.add(instance.attributes.instance_family)
    }
    if (instance.attributes?.instance_size) {
      uniqueValues.sizes.add(instance.attributes.instance_size)
    }
  })

  if (uniqueValues.sizes.size > 0) {
    requirements.push({
      key: 'InstanceSize',
      operator: 'In',
      values: Array.from(uniqueValues.sizes),
    })
  }

  if (uniqueValues.architectures.size > 0) {
    requirements.push({
      key: 'Arch',
      operator: 'In',
      values: Array.from(uniqueValues.architectures),
    })
  }

  if (uniqueValues.families.size > 0) {
    requirements.push({
      key: 'InstanceFamily',
      operator: 'In',
      values: Array.from(uniqueValues.families),
    })
  }

  return requirements
}
