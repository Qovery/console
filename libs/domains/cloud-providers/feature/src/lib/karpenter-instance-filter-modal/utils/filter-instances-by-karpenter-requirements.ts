import { type ClusterInstanceTypeResponseListResultsInner } from 'qovery-typescript-axios'
import { type KarpenterData } from '@qovery/shared/interfaces'

export function filterInstancesByKarpenterRequirements(
  allInstances: ClusterInstanceTypeResponseListResultsInner[],
  karpenterData: Omit<KarpenterData, 'disk_size_in_gib' | 'enabled' | 'spot_enabled'>
): ClusterInstanceTypeResponseListResultsInner[] {
  const karpenterRequirements = karpenterData.qovery_node_pools?.requirements || []

  return allInstances.filter((instance) => {
    const instanceSize = instance.attributes?.instance_size || ''
    const InstanceFamily = instance.attributes?.instance_family || ''
    const instanceArchitecture = instance.architecture || ''

    return karpenterRequirements.every((requirement) => {
      if (requirement.key === 'InstanceSize') {
        return requirement.values.includes(instanceSize)
      }

      if (requirement.key === 'Arch') {
        // Fix inconsistency between uppercase and lowercase with architecture during the migration
        return requirement.values.map((v) => v.toUpperCase()).includes(instanceArchitecture)
      }

      if (requirement.key === 'InstanceFamily') {
        return requirement.values.includes(InstanceFamily)
      }

      return false
    })
  })
}

export default filterInstancesByKarpenterRequirements
