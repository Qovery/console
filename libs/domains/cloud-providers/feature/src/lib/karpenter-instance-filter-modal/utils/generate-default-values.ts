import { type ClusterInstanceTypeResponseListResultsInner } from 'qovery-typescript-axios'
import { type KarpenterInstanceFormProps } from '../karpenter-instance-filter-modal'

export function generateDefaultValues(
  instances: ClusterInstanceTypeResponseListResultsInner[]
): Omit<KarpenterInstanceFormProps, 'default_service_architecture'> {
  // Get unique architectures
  const architectures = [...new Set(instances.map((i) => i.architecture))]

  // Get unique sizes without sorting
  const sizes = [...new Set(instances.map((i) => i.attributes?.instance_size ?? ''))].filter((i) => i !== '')

  // Group instance families by raw instance_category
  const categories: Record<string, string[]> = {}
  instances.forEach((instance) => {
    const family = instance.attributes?.instance_family || ''
    const category = instance.attributes?.instance_category || ''

    if (!categories[category]) {
      categories[category] = []
    }

    if (!categories[category].includes(family)) {
      categories[category].push(family)
    }
  })

  return {
    ARM64: architectures.includes('ARM64'),
    AMD64: architectures.includes('AMD64'),
    categories,
    sizes,
  }
}

export default generateDefaultValues
