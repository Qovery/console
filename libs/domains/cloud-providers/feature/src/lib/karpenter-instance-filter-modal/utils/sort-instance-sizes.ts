type BasicSize = 'small' | 'medium' | 'large' | 'xlarge'
type SizePriority = Record<BasicSize, number>

// Define type for the sort key tuple
type SortKey = [priority: number, multiplier: number]

export function sortInstanceSizes(sizes: string[]): string[] {
  const sizePriority: SizePriority = {
    small: 1,
    medium: 2,
    large: 3,
    xlarge: 4,
  }

  function getSortKey(size: string): SortKey {
    if (isBasicSize(size)) {
      return [sizePriority[size], 0]
    }

    // Handle xlarge variants
    if (size.includes('xlarge')) {
      // Extract number before 'xlarge'
      const number = size.replace('xlarge', '')
      // Convert empty string to 1 (for just 'xlarge')
      const multiplier = number ? parseInt(number) : 1
      return [5, multiplier]
    }

    return [0, 0]
  }

  function isBasicSize(size: string): size is BasicSize {
    return size in sizePriority
  }

  return sizes.sort((a, b) => {
    const [priorityA, multiplierA] = getSortKey(a)
    const [priorityB, multiplierB] = getSortKey(b)

    if (priorityA !== priorityB) {
      return priorityA - priorityB
    }
    return multiplierA - multiplierB
  })
}

export default sortInstanceSizes
