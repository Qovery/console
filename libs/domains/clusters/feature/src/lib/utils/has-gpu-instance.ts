import { type Cluster } from 'qovery-typescript-axios'

export const hasGpuInstance = (cluster?: Cluster) => {
  const clusterFeatureKarpenter = cluster?.features?.find((feature) => feature.id === 'KARPENTER')
  if (!clusterFeatureKarpenter) return false
  return Boolean(
    typeof clusterFeatureKarpenter?.value_object?.value === 'object' &&
      'qovery_node_pools' in clusterFeatureKarpenter.value_object.value &&
      clusterFeatureKarpenter.value_object.value.qovery_node_pools.gpu_override
  )
}
