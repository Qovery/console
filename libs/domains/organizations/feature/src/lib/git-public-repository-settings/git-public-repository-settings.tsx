import { Controller, useFormContext } from 'react-hook-form'
import { ExternalLink, InputText } from '@qovery/shared/ui'
import { guessGitProvider } from '@qovery/shared/util-git'

export interface GitPublicRepositorySettingsProps {
  disabled?: boolean
  hideRootPath?: boolean
  urlRepository?: string
}

export function GitPublicRepositorySettings({
  disabled = false,
  urlRepository,
  hideRootPath,
}: GitPublicRepositorySettingsProps) {
  const { control, setValue } = useFormContext()

  const onRepositoryChange = (value: string) => {
    const provider = guessGitProvider(value)
    if (provider) {
      setValue('provider', provider)
    }
  }

  return (
    <>
      <Controller
        name="repository"
        control={control}
        rules={{
          required: 'Repository required',
          validate: (input) =>
            // eslint-disable-next-line no-useless-escape
            input.match(/^(http(s)?:\/\/)[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:\/?#[\]@!\$&'\(\)\*\+,;=.]+.git$/gm) !==
              null || 'URL must be valid, start with «http(s)://» and end with «.git»',
        }}
        render={({ field, fieldState: { error } }) => (
          <div>
            <InputText
              label="Public repository URL (.git)"
              name={field.name}
              onChange={(e) => {
                field.onChange(e)
                onRepositoryChange(e.target.value)
              }}
              value={field.value}
              error={error?.message}
              disabled={disabled}
            />
            {urlRepository && (
              <ExternalLink className="ml-3" size="xs" href={urlRepository}>
                Go to repository
              </ExternalLink>
            )}
          </div>
        )}
      />
      <Controller
        name="branch"
        control={control}
        rules={{
          required: 'Branch required',
        }}
        render={({ field, fieldState: { error } }) => (
          <InputText
            label="Branch"
            name={field.name}
            onChange={(event) => {
              event.target.value = event.target.value.trim()
              field.onChange(event)
            }}
            value={field.value}
            error={error?.message}
            disabled={disabled}
          />
        )}
      />
      {!hideRootPath && (
        <>
          <Controller
            name="root_path"
            control={control}
            defaultValue="/"
            rules={{
              required: 'Value required',
            }}
            render={({ field, fieldState: { error } }) => (
              <InputText
                label="Root service path"
                name={field.name}
                onChange={field.onChange}
                value={field.value}
                error={error?.message}
                disabled={disabled}
              />
            )}
          />
          <p className="ml-4 mt-1 text-xs text-neutral-350">
            Provide the path in the repository where the service is located
          </p>
        </>
      )}
    </>
  )
}

export default GitPublicRepositorySettings
