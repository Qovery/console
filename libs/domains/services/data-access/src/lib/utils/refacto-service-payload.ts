import {
  type ApplicationEditRequest,
  type ApplicationGitRepositoryRequest,
  DatabaseModeEnum,
  type HelmRequest,
  type HelmResponseAllOfSourceOneOf,
  type HelmResponseAllOfSourceOneOf1,
  type ServiceStorageStorageInner,
} from 'qovery-typescript-axios'
import { match } from 'ts-pattern'
import { isJobGitSource } from '@qovery/shared/enums'
import { type Application, type Container, type Database, type Helm, type Job } from '../domains-services-data-access'

/* 
TODO: all this following functions should be removed after the API refactoring and we need to clean it after the Services migration to React Query 
https://www.notion.so/qovery/API-improvements-b54ba305c2ee4e549eb002278c532c7f?pvs=4#7d71ea23b5fa44ca80c15a0c32ebd8da
*/

export function refactoApplication(application: Application) {
  // refacto because we can't send all git data
  if (application.git_repository) {
    application.git_repository = {
      url: application.git_repository.url,
      branch: application.git_repository.branch,
      root_path: application.git_repository.root_path,
      git_token_id: application.git_repository.git_token_id,
    }
  }

  // refacto to remove the id by storage
  if (application.storage) {
    application.storage =
      application.storage.length > 0
        ? application.storage.map((storage: ServiceStorageStorageInner) => ({
            mount_point: storage.mount_point,
            size: storage.size,
            type: storage.type,
            id: storage.id,
          }))
        : []
  }

  const applicationRequestPayload: ApplicationEditRequest = {
    name: application.name,
    storage: application.storage,
    cpu: application.cpu,
    git_repository: application.git_repository as ApplicationGitRepositoryRequest,
    build_mode: application.build_mode,
    description: application.description || undefined,
    memory: application.memory,
    auto_preview: application.auto_preview,
    auto_deploy: application.auto_deploy,
    ports: application.ports,
    dockerfile_path: application.dockerfile_path || undefined,
    healthchecks: application.healthchecks ?? {},
    buildpack_language: application.buildpack_language,
    max_running_instances: application.max_running_instances,
    min_running_instances: application.min_running_instances,
    entrypoint: application.entrypoint,
    arguments: application.arguments,
  }

  return applicationRequestPayload
}

export function refactoContainer(container: Container) {
  const containerRequestPayload = {
    name: container.name || '',
    description: container.description || '',
    storage: container.storage,
    ports: container.ports,
    cpu: container.cpu,
    memory: container.memory,
    max_running_instances: container.max_running_instances,
    min_running_instances: container.min_running_instances,
    registry_id: container.registry?.id || '',
    image_name: container.image_name || '',
    tag: container.tag || '',
    arguments: container.arguments,
    entrypoint: container.entrypoint,
    auto_preview: container.auto_preview,
    auto_deploy: container.auto_deploy,
    healthchecks: container.healthchecks,
  }

  return containerRequestPayload
}

export function refactoJob(job: Job) {
  const jobRequest = {
    name: job.name,
    description: job.description,
    cpu: job.cpu,
    memory: job.memory,
    auto_preview: job.auto_preview,
    auto_deploy: job.auto_deploy,
    max_duration_seconds: job.max_duration_seconds,
    port: job.port,
    max_nb_restart: job.max_nb_restart,
    schedule: job.schedule,
    healthchecks: job.healthchecks ?? {},
    source: {},
  }

  if (isJobGitSource(job.source)) {
    jobRequest.source = {
      docker: {
        dockerfile_path: job.source.docker?.dockerfile_path,
        git_repository: {
          url: job.source.docker?.git_repository?.url || '',
          branch: job.source.docker?.git_repository?.branch,
          root_path: job.source.docker?.git_repository?.root_path,
          git_token_id: job.source.docker?.git_repository?.git_token_id,
        },
      },
    }
  } else {
    jobRequest.source = {
      image: {
        registry_id: job.source?.image?.registry_id,
        image_name: job.source?.image?.image_name,
        tag: job.source?.image?.tag,
      },
    }
  }

  return jobRequest
}

export function refactoDatabase(database: Database) {
  const databaseRequestPayload = {
    name: database.name,
    description: database.description,
    version: database.version,
    accessibility: database.accessibility,
    cpu: database.cpu,
    memory: database.memory,
    storage: database.storage,
    instance_type: database.instance_type,
  }

  if (database.mode === DatabaseModeEnum.MANAGED) databaseRequestPayload.version = database.version

  return databaseRequestPayload
}

export function refactoHelm(helm: Helm) {
  const sourceProvider = (helm?.source as HelmResponseAllOfSourceOneOf).git ? 'GIT' : 'HELM_REPOSITORY'

  const source = match(sourceProvider)
    .with('GIT', () => {
      const gitRepository = (helm?.source as HelmResponseAllOfSourceOneOf).git?.git_repository
      return {
        git_repository: {
          url: gitRepository?.url ?? '',
          branch: gitRepository?.branch,
          root_path: gitRepository?.root_path,
          git_token_id: gitRepository?.git_token_id,
        },
      }
    })
    .with('HELM_REPOSITORY', () => {
      const helmRepository = (helm?.source as HelmResponseAllOfSourceOneOf1).repository
      return {
        helm_repository: {
          repository: helmRepository?.repository?.id,
          chart_name: helmRepository?.chart_name,
          chart_version: helmRepository?.chart_version,
        },
      }
    })
    .exhaustive()

  const helmRequestPayload: HelmRequest = {
    name: helm.name,
    description: helm.description,
    auto_preview: helm.auto_preview,
    auto_deploy: helm.auto_deploy,
    ports: helm.ports,
    allow_cluster_wide_resources: helm.allow_cluster_wide_resources,
    arguments: helm.arguments,
    values_override: helm.values_override,
    source,
  }

  return helmRequestPayload
}
