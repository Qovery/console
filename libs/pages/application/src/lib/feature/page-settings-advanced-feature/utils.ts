import { ApplicationAdvancedSettings } from 'qovery-typescript-axios'
import { ApplicationEntity } from '@console/shared/interfaces'

export function initFormValues(keys: string[], application: ApplicationEntity): { [key: string]: string } {
  const values: { [key: string]: string } = {}
  keys.forEach((key) => {
    if (application.advanced_settings?.current_settings) {
      values[key] =
        application.advanced_settings.current_settings[key as keyof ApplicationAdvancedSettings]?.toString() || ''
    }
  })

  return values
}
