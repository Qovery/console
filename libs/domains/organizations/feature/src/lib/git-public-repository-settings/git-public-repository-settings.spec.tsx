import { wrapWithReactHookForm } from '__tests__/utils/wrap-with-react-hook-form'
import { renderWithProviders } from '@qovery/shared/util-tests'
import { GitPublicRepositorySettings } from './git-public-repository-settings'

describe('GitPublicRepositorySettings', () => {
  it('should match snapshot', () => {
    const { baseElement } = renderWithProviders(
      wrapWithReactHookForm(<GitPublicRepositorySettings urlRepository="https://github.com/qovery/console" />)
    )
    expect(baseElement).toMatchSnapshot()
  })
})
