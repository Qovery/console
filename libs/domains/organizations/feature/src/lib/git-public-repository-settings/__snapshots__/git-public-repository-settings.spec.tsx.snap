// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitPublicRepositorySettings should match snapshot 1`] = `
<body>
  <div>
    <div>
      <div
        class=""
        data-testid="input-text-wrapper"
      >
        <div
          class="relative"
        >
          <div
            aria-label="input-container"
            class="input     "
          >
            <div
              class=""
            >
              <label
                class="translate-y-2 text-sm"
                for="Public repository URL (.git)"
              >
                Public repository URL (.git)
              </label>
              <input
                class="input__value "
                data-testid="input-text"
                id="Public repository URL (.git)"
                name="repository"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <a
        class="cursor-pointer transition duration-100 font-medium inline-flex items-center gap-1 text-sky-500 hover:text-sky-600 text-xs ml-3"
        href="https://github.com/qovery/console"
        rel="noopener noreferrer"
        target="_blank"
      >
        Go to repository
        <i
          aria-hidden="true"
          class="fa-solid fa-arrow-up-right-from-square text-2xs leading-4"
        />
      </a>
    </div>
    <div
      class=""
      data-testid="input-text-wrapper"
    >
      <div
        class="relative"
      >
        <div
          aria-label="input-container"
          class="input     "
        >
          <div
            class=""
          >
            <label
              class="translate-y-2 text-sm"
              for="Branch"
            >
              Branch
            </label>
            <input
              class="input__value "
              data-testid="input-text"
              id="Branch"
              name="branch"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class=""
      data-testid="input-text-wrapper"
    >
      <div
        class="relative"
      >
        <div
          aria-label="input-container"
          class="input    input--label-up "
        >
          <div
            class=""
          >
            <label
              class="translate-y-2 text-sm"
              for="Root service path"
            >
              Root service path
            </label>
            <input
              class="input__value "
              data-testid="input-text"
              id="Root service path"
              name="root_path"
              type="text"
              value="/"
            />
          </div>
        </div>
      </div>
    </div>
    <p
      class="ml-4 mt-1 text-xs text-neutral-350"
    >
      Provide the path in the repository where the service is located
    </p>
  </div>
</body>
`;
